File: cpp_version/array_and_strings/most_frequent_char/most_frequent_char.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// most frequent char
// Write a function, mostFrequentChar, that takes in a string as an argument.
// The function should return the most frequent character of the string.
// If there are ties, return the character that appears earlier in the string.

// You can assume that the input string is non-empty.

char mostFrequentChar(std::string s) {
    // todo
    std::unordered_map<char, int> hash;
    
    char maxChar = '\0';

    return maxChar;
}

int main() {
    std::cout << mostFrequentChar("bookeeper") << std::endl;     // -> 'e'
    std::cout << mostFrequentChar("david") << std::endl;         // -> 'd'
    std::cout << mostFrequentChar("abby") << std::endl;          // -> 'b'
    std::cout << mostFrequentChar("mississippi") << std::endl;   // -> 'i'
    std::cout << mostFrequentChar("potato") << std::endl;        // -> 'o'
    std::cout << mostFrequentChar("eleventennine") << std::endl; // -> 'e'
    std::cout << mostFrequentChar("riverbed") << std::endl;      // -> 'r'
    return 0;
}

File: javascript_version/array_and_strings/most_frequent_char/most_frequent_char.js
// most frequent char
// Write a function, mostFrequentChar, that takes in a string as an argument. 
// The function should return the most frequent character of the string. 
// If there are ties, return the character that appears earlier in the string.

// You can assume that the input string is non-empty.

const mostFrequentChar = (s) => {
    let hash = new Object();
    for (char of s) {
        hash[char] = (hash[char] || 0) + 1;
    }
    if (Object.values(hash).length === 0) { return null; }
    return Object.keys(hash).reduce((a, b) => hash[a] >= hash[b] ? a : b);
};




console.log(mostFrequentChar('bookeeper')); // -> 'e'
console.log(mostFrequentChar('david')); // -> 'd'
console.log(mostFrequentChar('abby')); // -> 'b'
console.log(mostFrequentChar('mississippi')); // -> 'i'
console.log(mostFrequentChar('potato')); // -> 'o'
console.log(mostFrequentChar('eleventennine')); // -> 'e'
console.log(mostFrequentChar("riverbed")); // -> 'r'